### ISSUE TITLE
Initial end-to-end integration with ARS

### ISSUE THREAD

**Issue by dkoslicki** [RTXteam/RTX#867]
From [this issue](https://github.com/NCATSTranslator/TranslatorArchitecture/issues/12#issue-645586070), we need to check:

- [x] Are we integrated with the ARS? Status check (@edeutsch)
- [x] Do we have a QueryGraphInterpreter that satisfies the requirement of 2 KP's
- [x] We will need to provide a QG that corresponds to the above

Due date is July 8th, 2020.

May want to make a PR to the ARS like BTE did [here](https://ncatstranslator.slack.com/archives/C014UNWRGSC/p1593551942008100)

**Comment by edeutsch** [comment]
> Are we integrated with the ARS? Status check (@edeutsch)

basically, yes, I would say. I have sent proposed queries we can answer to Mark and I they are working on the plumbing

> Do we have a QueryGraphInterpreter that satisfies the requirement of 2 KP's

I sent them this query_graph (among others)
```
  "query_graph": {
    "edges": [
      {
        "id": "e00",
        "source_id": "n00",
        "target_id": "n01",
      },
      {
        "id": "e01",
        "source_id": "n01",
        "target_id": "n02",
        "type": "physically_interacts_with"
      }
    ],
    "nodes": [
      {
        "curie": "DOID:14330",
        "id": "n00",
      },
      {
        "id": "n01",
        "is_set": true,
        "type": "protein"
      },
      {
        "id": "n02",
        "type": "chemical_substance"
      }
    ]
  }
```

This consults only one KP, ARAX/KG1. We would need a more advanced DSL program to consult more than one KP. Unless we decided to count NGD as another KP, in which case that is easy to add.

> We will need to provide a QG that corresponds to the above

That was my most complex QG given to them. It wheels in Jaccard, but nothing more complex than that.

The QueryGraphInterpreter needs to be made smarter to allow for different DSL based on the provided types. That's on my list.

I'm thinking coming up with a case that hits KG1 or KG2 plus COHD could be our two KPs.

**Comment by dkoslicki** [comment]
> I'm thinking coming up with a case that hits KG1 or KG2 plus COHD could be our two KPs.

We should also come up with cases that hit COHD and BTE, since it’s preferred that we hit at least two KP’s that are “not our own.” Especially since KG2 isn’t an “official KP” yet.

> and I they are working on the plumbing

Potential typo leads to confusion: so who’s working on the plumbing? I.e. are we blocking on them for the integration with the ARS? Or do we have more work to do on this front?

**Comment by edeutsch** [comment]
sorry about the typo, ignore the I. They are working on the plumbing. I have sent them example query_graphs that they can test with. But they do not really tick the 'two KPs' box. I think we need to clarify which query_graph template will trigger which DSL that hits two KPs

**Comment by dkoslicki** [comment]
Ok, we can do that (make templates) during the mini-hackathon later today

**Comment by dkoslicki** [comment]
Note: the above DSL has a couple syntax issues (commas). `/beta` wants it to look like:
```
{
    "edges": [
      {
        "id": "e00",
        "source_id": "n00",
        "target_id": "n01",
      },
      {
        "id": "e01",
        "source_id": "n01",
        "target_id": "n02",
        "type": "physically_interacts_with"
      }
    ],
    "nodes": [
      {
        "curie": "DOID:14330",
        "id": "n00",
      },
      {
        "id": "n01",
        "is_set": true,
        "type": "protein"
      },
      {
        "id": "n02",
        "type": "chemical_substance"
      }
    ]
  }
```

**Comment by edeutsch** [comment]
I'm guilty of poorly hand-editing the JSON in haste. Here is a version without the extra commas:
```
{
    "edges": [
      {
        "id": "e00",
        "source_id": "n00",
        "target_id": "n01"
      },
      {
        "id": "e01",
        "source_id": "n01",
        "target_id": "n02",
        "type": "physically_interacts_with"
      }
    ],
    "nodes": [
      {
        "curie": "DOID:14330",
        "id": "n00"
      },
      {
        "id": "n01",
        "is_set": true,
        "type": "protein"
      },
      {
        "id": "n02",
        "type": "chemical_substance"
      }
    ]
  }
```

**Comment by dkoslicki** [comment]
Work on this will be happening in the branch 867

**Comment by dkoslicki** [comment]
@edeutsch So here's an example of a DSL for a classic 2-hop node query meant specifically for questions of the type: `(:disease)--(:anything)-(:chemical_substance)` (i.e. start node is a disease, end node is a chemical_substance).
Note that this will be slow, slow, slow (due to API calls to COHD and NGD). But in my testing (and with the modified ranking algo in branch 867) seems to give somewhat good results. Will need to tweak the ranking algo in #783 to get the numbers a bit higher...
```
"add_qnode(name=DOID:11830, id=n00, type=disease)",
"add_qnode(type=protein, is_set=true, id=n01)",
"add_qnode(type=chemical_substance, id=n02)",
"add_qedge(source_id=n00, target_id=n01, id=e00)",
"add_qedge(source_id=n01, target_id=n02, id=e01)",
"expand(edge_id=e00, kp=ARAX/KG2)",
#"expand(edge_id=e00, kp=BTE)",  #<---- commented out until we figure out how to coalesce nodes
"expand(edge_id=e01, kp=ARAX/KG2)",
"overlay(action=overlay_clinical_info, paired_concept_frequency=true, source_qnode_id=n00, target_qnode_id=n02, virtual_relation_label=C1)",
"overlay(action=overlay_clinical_info, observed_expected_ratio=true, source_qnode_id=n00, target_qnode_id=n02, virtual_relation_label=C2)",
"overlay(action=overlay_clinical_info, chi_square=true, source_qnode_id=n00, target_qnode_id=n02, virtual_relation_label=C3)",
"overlay(action=predict_drug_treats_disease, source_qnode_id=n02, target_qnode_id=n00, virtual_relation_label=P1)",
"overlay(action=compute_ngd)",  #<--- could just substitute this for a virtual edge between the drug and disease for speed considerations
"resultify(ignore_edge_direction=true)",
"filter_results(action=limit_number_of_results, max_results=50)",
```
This basically amounts to "hit it with every clinical info we know about, use the ML model, and use NGD, then sort out the rest from there with the ranker and limiting the results"

**Comment by dkoslicki** [comment]
I've reached out to Casey Ta to see if we can't speed up the COHD calls, otherwise we might need to interleave some filtering in-between each `overlay_clinical_info` call...

**Comment by dkoslicki** [comment]
Just to check up with this: @amykglen and @edeutsch is the plan to get the node normalizer implemented and merged to master before Thursday so that we can use both BTE and COHD in our demo example query? Or is the node normalizer not quite ready and I need to come up with other query graph interpreter templates that don’t use BTE (since the curies it returns don’t play well with COHD and other ARAXi features)?

**Comment by edeutsch** [comment]
From my perspective the latest node_normalizer.py is in master and the back end is ready. The interface could still use some work, but it ready for most use. Let me know what is wrong or what you need and I'll make it work.

There's also a branch called `node_normalization` that maybe you're referring to? I don't know anything about that. Up to @amykglen to fill in on that.

**Comment by amykglen** [comment]
yeah - the `node-normalization` branch contains all the code that's actually using the new `NodeSynonymizer` and doing deduplication.

merging it to `master` before Thursday seems reasonable to me - the status is that the `NodeSynonymizer` is functioning well (the interface improvements can be adopted as they're rolled out) and the only remaining issues are the two failing `test_ARAX_workflows.py` tests written up in #885 and #887. not sure how difficult those will be, but once they're resolved, everything is good to go to merge into `master`.

so I suppose it's somewhat up to you @dkoslicki, depending on if you think #885/#887 can be addressed by then (or if you think it's ok to merge the code without fixes for those implemented).

**Comment by dkoslicki** [comment]
@amykglen @edeutsch Tomorrow is booked for me, but let’s plan on this: I will address #885 and #887 on Wednesday (or late Tuesday) and then we can merge `node-normalization` to master.

In the mean time: @amykglen can you verify that calling BTE works with other ARAXi commands (like NGD and COHD)? I.e. since on Thursday we need to reach out to two different KP’s, it’s critical that BTE will work with these commands. If note, please open issues and assign me to them. A simple “call BTE and then overlay some clinical info” would be sufficient to test

**Comment by amykglen** [comment]
sounds good to me! and yes - I'll test that out and create issues for you as needed.

**Comment by amykglen** [comment]
fyi, @dkoslicki - I confirmed that NGD, COHD (all 3 functionalities), Jaccard, and add_node_pmids all work with BTE queries in the `node-normalization` branch. (COHD seemed to have pretty low rates of being able to calculate a non-NaN value, but I verified that at least some were obtained.)

ran into some issues with:

- predict_drug_treats_disease: created issue #892
- FET: created issue #893

**Comment by dkoslicki** [comment]
Merging branch 867 into node-normalizer seems to have not caused any issues, but was not able to completely run all tests (may have crashed COHD as it's now refusing my queries with "Max retries exceeded" and a bunch of "500 internal server" errors).

So @amykglen while I am going to merge `master` into `node-normalizer`, we may have to wait until COHD comes back online before we can test if this merger broke anything

**Comment by dkoslicki** [comment]
Looks like all non-FET tests are fine with the merger, so I'll go ahead and push this since we are already aware of this in #893

**Comment by dkoslicki** [comment]
Ok @edeutsch, I've merged `node-normalization` into master and added our example required for tomorrow [here](https://github.com/RTXteam/RTX/blob/master/code/ARAX/ARAXQuery/ARAX_query_graph_interpreter_templates.yaml#L92-L108). Let me know when you have the new `ARAX_query_graph_interpreter.py` up so we can test the following JSON to see if it matches with the template and is handled properly:
```
{
   "edges": [
      {
         "id": "e00",
         "source_id": "n00",
         "target_id": "n01",
         "type": null
      }
   ],
   "nodes": [
      {
         "id": "n00",
         "curie": "DOID:11830",
         "type": "disease"
      },
      {
         "id": "n01",
         "type": "chemical_substance"
      }
   ]
}
```
If you can also roll this out to `/beta` or some other endpoint, that would also be convenient (so I can copy-paste that JSON for the demo).

**Comment by amykglen** [comment]
fyi @dkoslicki - for query templates that expand the same edge with multiple KPs, you might want to use `continue_if_no_results=true`, like:
```
      - expand(edge_id=e00, kp=ARAX/KG2, continue_if_no_results=true)
      - expand(edge_id=e00, kp=BTE, continue_if_no_results=true)
```
because otherwise, if, for example, it happened that no paths were found in KG2, execution would halt there, even if BTE might have some results for that edge...

**Comment by edeutsch** [comment]
First time I tried it seemed to work okay for a while, but then timed out after an hour at:

2020-07-09 07:03:11.379571 DEBUG:  Computing observed expected ratios.
2020-07-09 07:03:11.379578 INFO:  Overlaying observed expected ratios utilizing Columbia Open Health Data. This calls an external knowledge provider and may take a while

**Comment by dkoslicki** [comment]
Thanks for pointing that out @amykglen ! Will change it when I have time (out of the house atm) but feel free to change in master too.

@edeutsch that's expected behavior given COHD is slow, and depending on the input disease, can result in many thousands of calls to COHD.

**Comment by amykglen** [comment]
cool - just changed it in `master`, @dkoslicki

**Comment by edeutsch** [comment]
Second time through it completed in 30 minutes.
I wonder if more feedback can be provided as queries are performed? Or is it one big batch with no opportunity for progress reporting?

**Comment by dkoslicki** [comment]
@edeutsch 
>I wonder if more feedback can be provided as queries are performed? Or is it one big batch with no opportunity for progress reporting?

Yes, I could do some more progress reporting, but it might result in thousands of messages being reported. It's one big batch and I _could_ report on each step in the batch.

**Comment by edeutsch** [comment]
Currently for NGD I do produce one message per term. Which can produce a lot of messages, but you do get feedback that things are happening.

I've been musing about adding a new logging type PROGRESS (in addition to DEBUG, INFO, WARNING, ERROR). PROGRESS would be displayed in the live progress window, but filtered out in the final logging list displayed with the nice colors in the Messages tab.

**Comment by dkoslicki** [comment]
Ah, I like that idea of having a PROGRESS logging option!

**Comment by dkoslicki** [comment]
@edeutsch Any update on the ARS integration?

**Comment by dkoslicki** [comment]
Note [this example](https://github.com/RTXteam/RTX/issues/867#issuecomment-655882124) no longer works on /beta (which has the most recent `master` code), so I've opened an issue #930 to see what went wrong

**Comment by edeutsch** [comment]
Regarding ARS integration: I looked for the PR example from BTE and could not find it any more. And I vaguely recall someone saying it was not quite what they wanted. I reached out to Mark W and he is putting together a set of instructions for doing what they want. He is going to get together with another unspecified group tomorrow to work through the instructions with them. And then we're next. He will reach out to me after he has refined the instructions after tomorrow, and then we will work through it to get us registered.

**Comment by dkoslicki** [comment]
@edeutsch [here](https://docs.google.com/document/d/1w54dlbep2BtmLorGaeJ86l1YpSgjqSt1Ra9Ye15hBzs/edit?usp=sharing) are the directions for integrating into the ARS

**Comment by edeutsch** [comment]
I have created the PR to integrate with the ARS. Mark says he will look it over and test:
https://github.com/NCATSTranslator/Relay/pull/19

**Comment by dkoslicki** [comment]
All done and good to close
